02-peer-protocol.md

```diff
@@ -49,11 +68,468 @@ transaction is confirmed are also not persistent - until you know the script
+contributions.
+
+1. type: 70 (`tx_complete`)
+2. data:
+    * [`channel_id`:`channel_id`]
+
+#### Requirements
+
+The nodes:
+  - MUST send this message in succession to conclude this protocol
+
+The receiving node:
+  - MUST use the negotiated inputs and outputs to construct a transaction
+  - MUST fail the negotiation if:
+    - the peer's total input satoshis is less than their outputs
+    - the peer's paid feerate does not meet or exceed the agreed `feerate`,
+      (based on the `minimum fee`).
+    - if is the *non-initiator*:
+      - the *initiator*'s fees do not cover the `common` fields
+    - there are more than 252 inputs
+    - there are more than 252 outputs
+    - the estimated weight of the tx is greater than 400,000 (`MAX_STANDARD_TX_WEIGHT`)
+
+#### Rationale
+
+To signal the conclusion of exchange of transaction inputs and outputs.
+
+Upon successful exchange of `tx_complete` messages, both nodes
+should construct the transaction and proceed to the next portion of the
+protocol. For channel establishment v2, exchanging commitment transactions.
+
+For the `minimum fee` calculation see [BOLT #3](03-transactions.md#calculating-fees-for-collaborative-transaction-construction).
+
+The maximum inputs and outputs are capped at 252. This effectively fixes
+the byte size of the input and output counts on the transaction to one (1).
+
+### The `tx_signatures` Message
+
+1. type: 71 (`tx_signatures`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`sha256`:`txid`]
+    * [`u16`:`num_witnesses`]
+    * [`num_witnesses*witness_stack`:`witness_stack`]
+
+1. subtype: `witness_stack`
+2. data:
+    * [`u16`:`num_input_witness`]
+    * [`num_input_witness*witness_element`:`witness_element`]
+
+1. subtype: `witness_element`
+2. data:
+    * [`u16`:`len`]
+    * [`len*byte`:`witness`]
+
+#### Requirements
+
+The sending node:
+  - if it has the lowest total satoshis contributed, as defined
+    by total `tx_add_input` values:
+    - MUST transmit their `tx_signatures` before the peer
+  - MUST order the `witness_stack`s by the `serial_id` of the input they
+    correspond to
+  - number of `witness_stack`s MUST equal the number of inputs they added
+
+The receiving node:
+  - MUST fail the negotiation if:
+    - if they have not sent a `tx_signatures` message and they are the
+      peer with the lowest total satoshis contributed, as defined by
+      `tx_add_input` values
+    - the message contains an empty `witness_stack`
+    - the number of `witness_stack`s does not equal the number of inputs
+      added by the sending node
+    - the `txid` does not match the txid of the transaction
+  - SHOULD apply the `witness`es to the transaction and broadcast it
+  - MUST reply with their `tx_signatures` if not already transmitted
+
+#### Rationale
+
+The peer with the lowest total of inputs amounts must transmit its
+`tx_signatures` first. If both peers have contributed equal input amounts,
```

Comment:
There is a griefing attack that results from this `tx_signatures` ordering.

Suppose you have configured your node to match funds from all `open_channel2`s up to a certain amount.

A malicious node can repeatedly initiate negotiations with your node, using UTXOs they don't own to construct the funding tx. The malicious node can ensure they have the larger contribution to the funding tx, forcing your node to send `tx_signatures` first. The malicious node then refuses to send their `tx_signatures`, forcing lockup of your funds until your node times out and double-spends them.

In this way, malicious node(s) can force your node to either lockup all its available funds or  repeatedly waste tx fees from double-spends, eventually consuming all your available funds.


Would it make more sense to require the opener to send `tx_signatures` first?  This would allow the acceptor to grief the opener, but the opener can easily go elsewhere if they get burned.

Reply:
> Suppose you have configured your node to match funds from all open_channel2s up to a certain amount.

That is a strong assumption, a node using that behavior knows that they are taking some risk and must implement some kind of DoS protection anyway. I wouldn't recommend such a strategy anyway.

> In this way, malicious node(s) can force your node to either lockup all its available funds or repeatedly waste tx fees from double-spends, eventually consuming all your available funds.

Not really, because the attacked node would just reuse the same utxos in another channel opening flow, which provides the double-spending part "for free" (no fees paid). The timeout for receiving `tx_signatures` can be very low, as there's no reason for an honest node to withhold their `tx_signatures` for a long time.

> Would it make more sense to require the opener to send tx_signatures first?

No, I think this would be worse in terms of protocol flexibility as this would remove the ability for an opener to batch funding transactions by coordinating multiple sessions in parallel. It's trivial to avoid the griefing attack you mention by just contributing 1 sat less than the opener (if you insist on matching their contribution), so I don't think we should be very concerned about this.

Reply:
> That is a strong assumption, a node using that behavior knows that they are taking some risk and must implement some kind of DoS protection anyway. I wouldn't recommend such a strategy anyway.

The exact policy isn't too important, since the attacker can defeat any of them.  For example, here's CLN's available policies:

- match -- Contribute policy_mod percent of their requested funds. Valid policy_mod values are 0 to 200. If this is a channel lease request, we match based on their requested funds. If it is not a channel lease request (and lease_only is false), then we match their funding amount. Note: any lease match less than 100 will likely fail, as clients will not accept a lease less than their request.
- available -- Contribute policy_mod percent of our available node wallet funds. Valid policy_mod values are 0 to 100.
- fixed -- Contributes a fixed policy_mod sats to v2 channel open requests.

For any of these, the attacker can keep doing `tx_add_input` to maintain a larger contribution.

But I am curious... How do you see dual funding being used without some fixed policy like these?  

> Not really, because the attacked node would just reuse the same utxos in another channel opening flow, which provides the double-spending part "for free" (no fees paid). The timeout for receiving `tx_signatures` can be very low, as there's no reason for an honest node to withhold their `tx_signatures` for a long time.

Right, we can productively double-spend, which helps against the attack.  But we still need to remember every pending channel using a UTXO until one of them confirms or we double spend.  If the attacker is blasting us with pending channels that will *never* confirm, it will eat up our memory/disk space until we crash or are forced to double-spend.

> No, I think this would be worse in terms of protocol flexibility as this would remove the ability for an opener to batch funding transactions by coordinating multiple sessions in parallel. It's trivial to avoid the griefing attack you mention by just contributing 1 sat less than the opener (if you insist on matching their contribution), so I don't think we should be very concerned about this.

I think you meant 1 sat *more* than the opener.

It's not so trivial.  The `tx_signatures` order is determined by value of inputs contributed, not by the `funding_sats` specified in `open_channel2`/`accept_channel2`.  So the attacker can keep doing `tx_add_input` to maintain a larger contribution (remember they have the entire UTXO set to pick from).

Reply:
> But that's something you have to do all the time, right? You always need to watch unconfirmed channels until they either confirm or are double-spent (or, if you didn't participate in funding, you can just forget them after a while). You will likely be running out of utxos to use in funding before memory becomes a problem, so I'm not sure you should be so concerned about this.

In the v1 flow, the acceptor can safely forget the pending channel at any point, since they didn't contribute any funds.  In v2, a dual-funding acceptor can't forget until an input in the pending channel is double spent.  Since the acceptor is attempting to productively double-spend their UTXOs, they may not actually run out -- they keep reusing the same UTXOs for each dual-funding request, consuming additional memory each time for the new pending channel.

> I'm not sure what is best, we could also make the opener sign first by default, with an optional TLV in `open_channel2` that would let them say "Hey, I'm trying to batch, please be a gent and send tx_signatures first this time"? And the non-initiator would contribute to only a handful of those per-day (or any restricting policy) to restrict their exposure. If this is used in the context of liquidity ads, since the non-initiator gets a fee in the end, it's probably a good enough incentive to take some risk and sign first? What do you think? Curious to know what @niftynei thinks about this as well.

It seems for simple dual funding, opener signing first is the simplest.  But if we're going to optimize for the 2-party case, the whole interactive-tx session is unnecessary anyway (we could just send all inputs/outputs once).

I cringe thinking about the complexity, but maybe something like PoDLEs are the ultimate solution here...

Reply:
From my understanding this issue of lack of atomicity of signatures exchanges, introducing DoS vectors or economic disequilibrium of the collaborative transaction has been already hit within the Discreet Log Contract specification effort. Analogously, two participants are contributing collateral UTXOs to the same off-chain contract state over multiple Lightning hops. One idea brought out to solve the issue was "Barrier Escrow", a third-party guaranteeing the release of participants signature scalars: https://suredbits.com/payment-points-and-barrier-escrows/. The trust model itself stays to be studied.
 
About PoDLEs, I think they could be a significant v2 channels coins griefing vector rate-limiting, in the sense you can share the non-honored UTXO commitments among LSP (let's say over gossips). I don't think this obliterate the risks for an individual LSP in face of an adversary able to renew its stock of griefing UTXOs in period of low-fees. 

Note, I don't think this is something we should aim for either v0.1 or v0.2 of the interactive transaction construction protocol, however that might be long-term set of solutions.